ocaml/xenopsd/xc/service.ml

```diff
@@ -164,3 +179,591 @@ let start_and_wait_for_readyness ~task ~service =
+   (optionally with a given value) If this doesn't happen in the timeout then
+   an exception is raised *)
+let wait_path ~pidalive ~task ~name ~domid ~xs ~ready_path ~timeout ~cancel _ =
+  let syslog_key = Printf.sprintf "%s-%d" name domid in
+  let watch = Watch.value_to_appear ready_path |> Watch.map (fun _ -> ()) in
+  Xenops_task.check_cancelling task ;
+  ( try
+      let (_ : bool) =
+        Cancel_utils.cancellable_watch cancel [watch] [] task ~xs ~timeout ()
+      in
+      ()
+    with Watch.Timeout _ ->
+      if pidalive name then
+        raise (Service_failed (name, "Timeout reached while starting daemon")) ;
+      raise (Service_failed (name, "Daemon exited unexpectedly"))
+  ) ;
+  debug "Daemon initialised: %s" syslog_key
+
+let pid_alive pid name =
+  match Forkhelpers.waitpid_nohang pid with
+  | 0, Unix.WEXITED 0 ->
+      true
+  | _, Unix.WEXITED n ->
+      error "%s: unexpected exit with code: %d" name n ;
+      false
+  | _, (Unix.WSIGNALED n | Unix.WSTOPPED n) ->
+      error "%s: unexpected signal: %d" name n ;
+      false
+
+module type DAEMONPIDPATH = sig
+  val name : string
+
+  val use_pidfile : bool
+
+  val pid_path : int -> string
+end
+
+module DaemonMgmt (D : DAEMONPIDPATH) = struct
+  module SignalMask = struct
+    module H = Hashtbl
+
+    type t = (int, bool) H.t
+
+    let create () = H.create 16
+
+    let set tbl key = H.replace tbl key true
+
+    let unset tbl key = H.remove tbl key
+
+    let has tbl key = H.mem tbl key
+  end
+
+  let signal_mask = SignalMask.create ()
+
+  let name = D.name
+
+  let pid_path = D.pid_path
+
+  let pid_path_signal domid = pid_path domid ^ "-signal"
+
+  let pidfile_path domid =
+    if D.use_pidfile then
+      Some
+        (Printf.sprintf "%s/%s-%d.pid" Device_common.var_run_xen_path D.name
+           domid
+        )
+    else
+      None
+
+  let pid ~xs domid =
+    try
+      match pidfile_path domid with
+      | Some path when Sys.file_exists path ->
+          let pid =
+            path |> Unixext.string_of_file |> String.trim |> int_of_string
+          in
+          Unixext.with_file path [Unix.O_RDONLY] 0 (fun fd ->
+              try
+                Unix.lockf fd Unix.F_TRLOCK 0 ;
+                (* we succeeded taking the lock: original process is dead.
+                 * some other process might've reused its pid *)
+                None
+              with Unix.Unix_error (Unix.EAGAIN, _, _) ->
+                (* cannot obtain lock: process is alive *)
+                Some pid
+          )
+      | _ ->
+          (* backward compatibility during update installation: only has
+             xenstore pid *)
+          let pid = xs.Xs.read (pid_path domid) in
+          Some (int_of_string pid)
+    with _ -> None
+
+  let is_running ~xs domid =
+    match pid ~xs domid with
+    | None ->
+        false
+    | Some p -> (
+      try Unix.kill p 0 ; (* This checks the existence of pid p *)
```

Comment:
The man page for kill(2) is missing from centos7 hosts (and CH), but on my dev machine seems to confirm this behaviour:

> If sig is 0, then no signal is sent, but existence and permission checks are still performed; this can be used to check  for the existence of a process ID or process group ID that the caller is permitted to signal.